Issue 435 citation depth

mattdahl opened the pull request:
This PR adds support for extracting "multiple-use" citations from a court opinion, i.e., extracting the number of times a given opinion cites another one. Currently you only have binary data on this. Cf. #435. This PR only contains logic for *extracting* and *saving* such citations, leaving the actual utilization of that data to future work (e.g., improving the search graph).

I've squashed my commits and rebased against master so this should be a clean diff. Below, see a summary of the changes made and a list of some outstanding problems.

Happy to discuss this and make improvements! @mlissner 

## Summary of changes
- In `models.py` I have created several new sub-classes of the base `Citation` class: `FullCitation`, `ShortformCitation`, `IdCitation` and `SupraCitation`. These classes are not backed in the database but rather help represent the conceptual (and textual) differences between the different kinds of citations we can now extract.
- Those objects are instantiated in the `get_citations()` function in `find_citations.py`. Basically, we loop through and parse the list of string tokens in an opinion, see whether a citation object can be generated from each token, and create such an object if so.
- Then, in the `get_citation_matches()` function in `match_citations.py`, we resolve those extracted citations to actual `Opinion` objects from the database. Matching full citations is done the same way as before (i.e., just looking them up in the DB), but the other citation types (i.e., "Id.", "supra," etc.) are resolved through a variety of different heuristics, since they necessarily can't be matched directly.
- Finally, we update the opinion document with this new data. First, using the `as_regex()` and `as_html()` methods of each citation object, we create meaningful HTML links in the opinion text to each kind of citation. (This logic already exists for full citations, but now all citation types are supported.) Second, we count how many times each matched opinion is cited in the document, and save that as a new field on the `OpinionsCited` model.
- This PR looks big in terms of lines of code, but most of that is additions to the tests and adding new fixtures. No modifications (except for semantic ones) were made to the existing test cases, so in theory no regressions have been introduced by these changes. I tried to add robust unit tests for all stages of the extracting/matching process described above.

## Some problems
- Speed. These changes necessarily slow down the parsing process. I have not done robust profiling, but on my local machine with 715 real opinions, running the `cl_find_citations` command is now about 27% slower. (0.89 seconds per opinion, as opposed to 0.70 seconds before.)
- Id./Ibid. citations are sometimes resolved incorrectly. This happens when the Id./Ibid. token is a reference to any document that is *not* an opinion. So if the text is citing a statute (or something), and then says "Id." thereafter, the reference there is really to the statute (which is something we don't collect), but the algorithm thinks that it's a reference to whatever the last cited *opinion* was. So we might erroneously overcount certain citations. I can't think of a way to avoid this problem, but I think that it's worth biting the bullet on this, because most of the time the Id./Ibid. *does* resolve correctly and that gives us a lot of helpful information. Furthermore, there is [research that suggests](https://doi.org/10.1093/pan/mps019) that methods of citation counting need not be 100% accurate to be meaningful, as even imprecise counting "results in a measure that is still very highly correlated with the actual number of citations."
- HTML links are not generated for citations that cross line breaks. We still count these citations, but it's hard to generate the proper HTML replacement for them without messing up the columnar appearance of an opinion, since we don't know where precisely the line break character should be placed in the generated HTML. In theory this is a solvable problem but it's too complex for me, so I just left it alone.

mlissner commented:
Re problems:

 - Speed is fine. We're getting a new server that'll be at least 10× faster on disk and who knows how much faster on CPU (it'll have 128 cores). 

 - Re the ibids and other types of citation, there are tools for identifying citations to the law, and we have an open issue about identifying these: https://github.com/freelawproject/courtlistener/issues/459. I wonder how much harder it'd be to get that in place before landing this (probably medium hard?)

 - HTML links at the end of lines are...a problem: https://github.com/freelawproject/courtlistener/issues/265. Did you find a way to just not linkify these? If so, maybe it fixes the bug above, or at least partially does so?

mattdahl commented:
Will talk more on Friday, but to respond briefly:
- Re: 2: That API looks great, and it would definitely help with this problem. Two possible issues: (1) It's too bad it's written in JS, so either we'd have to run a Node instance and interface that with Python somehow (annoying) or query their hosted version over HTTP (also annoying, and it looks like theirs is down?). (2) It looks like it only supports federal-level statutes, which is a good step forward, but may not solve the entirety of the problem here. (I suspect that many law citations are to state-level things, no? And those are probably all idiosyncratic, making them hard to catch.)
- Re: 3: I just wrote the regex to not match citations that have line break characters ("\n", "\r", etc.) in them. Those citations are still counted (because whitespace doesn't matter when we tokenize the opinion and parse it), but when it comes time to replace them with HTML links, the regex just won't match them if they cross a line break. I see that the current regex matches any whitespace character (`\s+`), which is probably causing the current bug. Of course, it's not really a bug, since ideally those citations *would* be linkified, just with the appropriate line break re-inserted. Perhaps if we created capturing groups in the regex for each whitespace option, and then checked whether what is captured is a line break character, and then used that knowledge to inform the HTML-building process? I sort of hate regex so I didn't want to wade into that, but it feels like something like that could work.

mlissner commented:
@mattdahl I just made a huge change to master that broke this PR. On top of that, we've recently been doing a lot of work on the citation finder in #1076. I wonder if you could see about getting this rebased on master and ready for my review one more time. I'm finally in a better place to review this once more. 

Did it have any outstanding issues before this came up? Do you recall?

Sorry for taking so long to get back to this.

mattdahl commented:
Sure, will rebase! I think things were working well before, but once I rebase I'll double-check and let you know what's up before you review.

mattdahl commented:
Okay, I've rebased this against master again, and the checks are passing, so I think this is ready for review @mlissner. Apologies for all the force pushing -- I wanted to keep the conceptual units of my original commits intact so I just replayed changes onto them one-by-one.

I also made some changes to address the issues discussed above (in February). On  No. 2, I ultimately found those regexes too brittle and too complicated to integrate, so I just added basic support for recognizing generic non-opinion citations instead -- using the section marker token (§) as a heuristic for identifying them. Recognizing these is important for not creating erroneous id/ibid citations, but right now no further parsing is done to try to do anything meaningful with the non-opinion citations we grab. Then, on No. 3, I realized that this problem was somewhat different than what I had described before. To make a long story short, I have now simply added regex backreferences to the generated inline HTML which will insert the line breaks at the appropriate places. This does not solve #265 but at least fixes the "problem" I had mentioned above.

I know that this is a big PR, so let me know if you want to talk through it or anything. My commits should represent individual coherent steps of development, which hopefully helps!

mlissner commented on cl/citations/tests.py:
```diff
@@ -355,39 +429,350 @@ def test_disambiguate_citations(self):
+             '<pre class="inline"></pre><span class="citation no-link"><span '
+             'class="antecedent">asdf, </span><span class="volume">1</span> '
+             '<span class="reporter">U.S.</span>, at <span class="page">2'
+             '</span></span><pre class="inline">. foobar</pre>'),
```
Does this example show a comma being added by our code after `U.S.`? That's probably an issue. Lawyers make big deals about commas.

mattdahl replied:
Yes, good catch, I will fix this so it only inserts a comma if there was one in the original text

mattdahl replied:
Fixed in b1e76cd

mlissner commented on cl/citations/tests.py:
```diff
@@ -355,39 +429,350 @@ def test_disambiguate_citations(self):
+            # First kind of supra citation (standard kind)
+            ('asdf, supra, at 2. foobar',
+             '<pre class="inline"></pre><span class="citation no-link"><span '
+             'class="antecedent">asdf,</span><span> supra</span><span>, at '
```
Seems like the span around `supra`, if it's worth keeping at all, should have  class? Larger point though: Why bother putting spans around the antecedents?

mattdahl replied:
This is just my bias for trying to enforce semantic markup everywhere. However, I agree that it's probably unnecessary here 😄

mattdahl replied:
Fixed (partially) in b1e76cd. It turns out that it's necessary to modify the general supra citation string at least a bit in order to prevent subsequent iterations of the regex/html replacement logic from trying to replace strings that have already been matched. So wrapping the antecedent guess in a `<span>` is one way to accomplish this; i.e., to mark that portion of the HTML as already replaced. Maybe there's an easier way to do this, but I don't think there's any harm in including that markup. (Unless we're trying to minimize bytes sent over the network... 😄)

mlissner replied:
OK. It looks fixed to me.

mlissner commented on cl/citations/tests.py:
```diff
@@ -355,39 +429,350 @@ def test_disambiguate_citations(self):
+
+            # Second kind of supra citation (with volume)
+            ('asdf, 123 supra, at 2. foo bar',
+             '<pre class="inline"></pre><span class="citation no-link"><span '
```
Weird to have a pre opening and closing itself here, with nothing inside?

mattdahl replied:
Yes, not sure what is happening, I will fix this

mattdahl replied:
It turns out that this behavior is intended, but just looks weird because the test string is too short. I added some more words at the beginning to make the test more realistic and to demonstrate what the full generated HTML is. b1e76cd

mlissner replied:
Ah ha.

mlissner commented on cl/citations/tests.py:
```diff
@@ -355,39 +429,350 @@ def test_disambiguate_citations(self):
+            ('asdf, Ibid. Lorem ipsum dolor sit amet',
+             '<pre class="inline">asdf, </pre><span class="citation no-link">'
+             'Ibid.<span class="after_tokens"> <span class="after_token">Lorem'
+             '</span> <span class="after_token">ipsum</span> <span class="'
```
Is there a reason to put each of the `after_token`s in their own `span`?

mattdahl replied:
No, this is sort of insane. Will fix

mattdahl replied:
Fixed in b1e76cd

mlissner commented on cl/citations/tests.py:
```diff
@@ -355,39 +429,350 @@ def test_disambiguate_citations(self):
+            ('asdf." Id., at 315.\n       Lorem ipsum dolor sit amet',
+             '<pre class="inline">asdf." </pre><span class="citation no-link">'
+             'Id.,<span class="after_tokens"> <span class="after_token">at'
+             '</span> <span class="after_token">315.</span>\n       <span'
```
I'm confused why `315` shows up as an `after_token` here, but in the `supra`'s these show up with the `page` class?

mattdahl replied:
This is because for `id` and `ibid`'s I don't attempt to do any parsing of the adjacent tokens at all. Instead, I just grab the three `after_token`'s (whatever they are), and use them as an "anchor" for building a regex string that correctly matches the proper `id` or `ibid` reference in the HTML. And since `id` and `ibid` don't always have page numbers, I saw no reason to try parsing them specifically (since they won't help in the matching process, even if they were extracted).

For `supra`s, by contrast, those regex strings *are* built with page numbers as a specific component. So they are extracted for that.

mlissner commented on cl/cleanup/tests.py:
```diff
@@ -293,7 +293,7 @@ def test_tax_court_citation_extractor(self):
             ),
             (
-                """
+                u"""
```
I'm curious what inspired this?

mattdahl replied:
The root of the issue is that this test case contains ascii characters, so I think Python assumes that the whole thing is ascii encoded. This is a problem because one of the "citation" tokens I scan for is the unicode section marker (https://github.com/mattdahl/courtlistener/blob/issue-435-citation-depth/cl/citations/find_citations.py#L600). So without explicitly declaring *this* test string as a unicode string, I get a `UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 4: ordinal not in range(128)` error when I try to parse it.

Perhaps I am doing something wrong myself; for the life of me I've never really been able to figure out how text encoding really works, and this just seemed like the easiest way to make everything play nicely together.

mlissner commented on cl/search/migrations/0089_opinionscited_depth.py:
```diff
@@ -0,0 +1,24 @@
+                help_text=b"The number of times the cited opinion was cited in the citing opinion",
+            ),
+        ),
+    ]
```
Do you mind pulling this and your model.py tweaks since I landed them manually? (The migration was a lot more complex for reasons you probably don't care to know.)

mattdahl replied:
👍

mattdahl replied:
Fixed in a8d3cbb (reverting my commit). I then had to merge master to get the changes you landed. (I didn't want to rebase again now that we started the review.)

mlissner commented on cl/citations/find_citations.py:
```diff
@@ -601,27 +552,70 @@ def get_citations(
         text = get_visible_text(text)
```
Your work in this section is fantastic. Thank you for the clear comments.

mlissner commented on cl/citations/find_citations.py:
```diff
@@ -377,66 +229,156 @@ def add_defendant(citation, words):
+    antecedent_guess = words[supra_index - 1]
+    if antecedent_guess.isdigit():
+        volume = int(antecedent_guess)
+        antecedent_guess = words[supra_index - 2]
```
Am I wrong or will these negative indexes either crash or just plain do weird things if supra_index = 0 or 1?

mattdahl replied:
Yes, I think you're right, but my assumption was that `supra_index` would *never* be 0 or 1, since it would be impossible for a supra citation to appear at the beginning of a document. (Since by definition, "supra" would have to refer to something *prior*.) However, I know that sane input is never guaranteed, and it's easy to add a check for this, so I will fix these 👍

mlissner replied:
These are on my radar because I know our regular parser crashes for this reason. But it is more likely to crash because lots of reporters can be the first word, but yeah, protecting against this will probably save us trouble.

mattdahl replied:
Fixed in d024203

mlissner commented on cl/citations/find_citations.py:
```diff
@@ -377,66 +229,156 @@ def add_defendant(citation, words):
+    volume = None
+
+    # Get page
+    page = parse_page(words[supra_index + 2])
```
Should we do an `if page is None` check here, in case parsing the page fails? I guess it would in something like "supra at page 22", right?

mattdahl replied:
It doesn't matter if `page` is `None`; indeed, some valid supra citations do not include page numbers. The code uses the page number if it is successfully parsed and doesn't if it isn't. (This is in contrast to a regular, full citation, which *must* contain a page reference, and so we check there if it doesn't.)

mlissner commented on cl/citations/match_citations.py:
```diff
@@ -139,3 +146,110 @@ def match_citation(citation, citing_doc=None):
+
+
+def get_citation_matches(opinion, citations):
+    # A list of opinions, as matched to citations
```
Can you add a proper docstring here please?

mattdahl replied:
👍

mattdahl replied:
Fixed in c839569

mlissner commented:
A thought. Would it make life easier if NonOpinionCitation wasn't a subclass of Citation? I couldn't quite figure this out on my own, but I saw a couple places where we could do something like, `if isinstance(citation, Citation)`, and thus make our lives easier, except that NonOpinionCitations would get caught in that too. Maybe you have a better gut feeling on this.

mlissner commented on cl/citations/match_citations.py:
```diff
@@ -139,3 +146,110 @@ def match_citation(citation, citing_doc=None):
+                # already matched opinions.
+                if (
+                    strip_punct(citation.antecedent_guess)
+                    in cm.cluster.case_name_full
```
I think I'd recommend using `base_case_name` here, which is just a helper function that gives you the `case_name`, `case_name_full`, or `case_name_short` of the item, in that order. I'd be careful though about using case_name_full, because they can be obscenely long (they're also often missing, which is what `best_case_name` helps with). I think using the shorter case name as a first pass and then falling back to the full case name is probably smart. Another thought....is it likely or possible that we'll see a lot of examples like this:

> ...v. IBM Inc., supra at 23

In which case we'd pick up on "Inc," which would be a really common term? I wonder if we need a blacklist for things like this that are common in lots of case names. "State" and "States" comes to mind too. What are the odds you have "Johnson v. State" and "Hamilton v. State" in the same case? Fairly high, I'd wager.

mattdahl replied:
1) Sure, that makes sense. I wasn't actually sure which case name field was best to use. However, I don't see `base_case_name` anywhere -- is that the name of the function?
2) Yes, excellent point. For some reason, I made the (erroneous) assumption that the `antecedent_guess` for supras would never cause a match conflict. By creating a blacklist, do you mean blacklisting possible antecedents during the parsing process? (So, if we encounter a blacklisted word -- like state or inc. -- we can keep scanning in the hopes of finding a more unique one?)

mlissner replied:
1. Sorry, typo. Should be `best_case_name`. 

2. I was thinking of a hardcoded blacklist of words like "state" and "inc" or whatever, but I'm not at all married to it. We might have a list like this in the code base already if you poke around for something like "v.", it might turn up? Sorry, I forget. 

    I guess another approach is to try matching the antecedent as we do now and if it matches more than one prior citation, give up. Working our ways backwards past common words is probably helpful too, so maybe a combination of both approaches would be safest?

mattdahl replied:
1. Fixed in c839569
2. In the same commit, I also changed the supra matching logic to give up if an antecedent guess matches two possible candidates, per your second suggestion. I think the blacklist idea is interesting, but something to add in a future PR? (Since this fix should at least avoid any problems of false matching, even if it doesn't take full advantage of all the information available.)

mlissner replied:
Yeah, seems OK for now. I guess we'll see when we get to real data exactly how much more refinement this needs.

mlissner commented on cl/citations/match_citations.py:
```diff
@@ -139,3 +146,110 @@ def match_citation(citation, citing_doc=None):
+                        refined_candidates.append(cm)
+
+                if len(refined_candidates) == 1:
+                    matched_opinion = refined_candidates[
```
Black did a nasty formatting here. Can you put your comment before the line so black does better?

mattdahl replied:
👍

mattdahl replied:
Fixed in c839569

mlissner commented on cl/citations/match_citations.py:
```diff
@@ -139,3 +146,110 @@ def match_citation(citation, citing_doc=None):
+        # If the citation is to a non-opinion document, we currently cannot
+        # match these.
+        elif isinstance(citation, NonopinionCitation):
+            pass
```
I think I'd move these two to the top. Start with the easy cases.

mattdahl replied:
👍

mattdahl replied:
Fixed in c839569

mlissner replied:
Should this first one be a `continue` instead of a `pass`?

mattdahl replied:
I don't think so, because I still want to trigger the logic at line 258 that toggles the `was_matched` variable appropriately (which is still within the for loop). I remember I used `continue` before but then encountered a bug, and I think switching to `pass` here fixed that.

mlissner commented on cl/citations/match_citations.py:
```diff
@@ -139,3 +146,110 @@ def match_citation(citation, citing_doc=None):
+        # was successful).
+        elif isinstance(citation, IdCitation):
+            if was_matched:
+                matched_opinion = citation_matches[-1]
```
These negative indexes are gonna bite us. Will this crash if the first citation we find is an `IdCitation`? I haven't done so carefully, but I suspect a careful pass over the PR looking at negative indexes would find some other instances too.

mattdahl replied:
It shouldn't crash here, because `was_matched` is initialized to `False`. So even if we find an `IdCitation` first, the code knows there's nothing in the `citation_matches` array to try to grab. However, I'll add a test case for this and review the negative indexing logic elsewhere.

mlissner replied:
1. Good point. 

2. Thanks.

mattdahl replied:
Fixed in c839569. This was the only instance of us doing negative indexing in this file; I addressed other possible problems in d024203.

mlissner commented:
OK, that's it for my review. Overall, this is fairly stunning. One of the bigger and more impressive PRs I've seen, especially since you made so much of it without contacting me very much. I think there's two areas of concern for me that my above comments might touch on, but which I think are worth calling out.

These feel like yellow flags:

1. Tests - Some of the tests surprised me. It seemed like the HTML was weird.

1. Regexes. Honestly, I skipped a careful review of the regexes because they're impossible to review.  I think if you feel comfortable with them, I do too, but if you don't, then one of us should dig into them deeper until somebody does.

1. False matches. It seems like you've made conservative choices pretty much throughout, which is pretty great, but I know from when we launched the first version of this that we had to really look at how it was doing in the real world and see what we discovered. I'm almost sure we're going to have to take a second pass at this once we see what real world bugs come out of it, but my concern is that I don't expect either of us will find those bugs since neither of us is (I assume) using the site for research. How to find these?

Thanks for the great and enjoyable read. This is going to be a great feature. One final question: Would you want to do a blog post on this?

mattdahl commented:
Thanks for the kind words. I've made some changes and tried to reference all your comments above individually. But, to also respond to your summary remarks:

1. Tests should be fixed, per comments above. Also added a few new cases for some things.
2. I did spend a lot of time trying to figure out these regexes, so I think I feel pretty confident in them. Of course, knock on wood...
3. Yes, I tried to make conservative choices (i.e., erring on the side of rejecting matches if there's ambiguity). I agree, however, that no doubt bugs will pop up in the wild that I have not anticipated. Once this is deployed, though, I actually do want to download some bulk data on the citation depths, so while I might not be using the site actively for research I will have a vested interest in making sure that things look okay on production. I'm also willing to continue to help out if problems with these changes crop up.
4. I took your advice about making the `NonopinionCitation` class a top-level object: 2d4e8ac. I do think this simplifies a few if statements.
5. Finally, I can certainly write a blog post about this if you want!

mlissner commented on cl/citations/match_citations.py:
```diff
@@ -139,3 +147,120 @@ def match_citation(citation, citing_doc=None):
+                # look alike, matches using this heuristic may not be unique.
+                # If no match, or more than one match, is found, then the supra
+                # reference is effectively dropped.
+                if strip_punct(citation.antecedent_guess) in best_case_name(
```
Seems like this punctuation stripping could be moved to the top? In a longer document, we're going to be running dozens of times. It'd also simplify the code, I think.

mattdahl replied:
Do you mean the top of the master `for` loop, i.e., around line 165? I can't put it there, unfortunately, because not all subclasses of `Citation` contain an `antecedent_guess` to be stripped. (So it has to go within each case of the appropriate `if`/`elif` blocks, at least. Though that's only twice.) I could separate the result of the stripping into its own variable, though, so that it isn't happening at the same time we're comparing it to the `best_case_name`?

mlissner replied:
That seems good, yeah.

mattdahl replied:
Fixed in a67cef7

mlissner commented:
OK, just a few more smallish things. I think. 

All the general topics sound good. For the blog post, the starting point is https://github.com/freelawproject/free.law/. I think the readme will get you going there, but if you get stuck, let me know. Looks like this is about ready to land.
